Add support for getting current timestamp in nano second precision.

higuoxing opened the pull request:
This patch adds support for getting current timestamp in nano second precision by using clock_gettime() [1]. I didn't have a Windows machine so I didn't implement it for Windows. Feedbacks are welcome.

[1] https://man7.org/linux/man-pages/man2/clock_gettime.2.html

fvannee commented on src/timestamp9--1.0.1--1.1.0.sql:
```diff
@@ -0,0 +1,3 @@
+CREATE FUNCTION timestamp9_nsnow() RETURNS timestamp9 AS
+'$libdir/timestamp9'
+LANGUAGE c IMMUTABLE STRICT PARALLEL SAFE LEAKPROOF;
```
This function as it is implemented is not `IMMUTABLE`, but it's `VOLATILE`. Normally, the Postgres `now()/current_date` etc. functions are neither of these, but are `STABLE`, which means that they return the same value across different calls in the same statement. I think any 'now' function that timestamp9 provides should be STABLE too. This means it likely requires a different implementation though.

higuoxing replied:
Thanks for your comments! Sounds reasonable, I'm will revise it.

higuoxing replied:
I'm using two static variables to 'cache' the timestamp inside a transaction. It's a little bit different from PostgreSQL's `now()`, since `now()` returns the first statement's timestamp inside a transaction to avoid extra cost of calling `GetCurrentTimestamp()`[1]. In this this approach, `timestamp9_nsnow()` returns the timestamp of the first `timestamp9_nsnow()` function call.

[1] https://github.com/postgres/postgres/blob/a8a44828a2e15c95a15549063205b779f7055f1c/src/backend/access/transam/xact.c#L2111

higuoxing committed:
Add support for getting current timestamp in nano second precision.

This patch adds support for getting current timestamp in nano second precision by using clock_gettime() [1].

[1] https://man7.org/linux/man-pages/man2/clock_gettime.2.html

fvannee commented on src/timestamp9.c:
```diff
@@ -647,3 +650,26 @@ Datum timestamp9_interval_mi(PG_FUNCTION_ARGS)
+{
+	timestamp9 result;
+	struct timespec tv;
+	static TimestampTz current_xact_timestamp;
```
Does this work in the case parallelism is enabled in queries? You can try using `force_parallel_mode` to test.

higuoxing replied:
Nice catch! It seems that the `current_xact_ts9_timestamp` cannot be synced to parallel workers in my implementation like `xactStartTimestamp`. Maybe we can mark it with `PARALLEL RESTRICTED`?

https://www.postgresql.org/docs/current/parallel-safety.html
